### PR TITLE
Feature/cplp 1524 user role collections

### PR DESCRIPTION
**Pull request by ntruchsess**
implements https://jira.catena-x.net/browse/CPLP-1524
corresponding devenv-PR: https://github.com/tractusx-team-portal-onboarding/devenv/pull/164

### CODE REVIEW

**Review by Phil91 — CHANGES_REQUESTED** [review]

File: src/administration/Administration.Service/Controllers/UserController.cs [hunk 1]
```diff
@@ -290,10 +357,11 @@ public async Task<CreatedAtRouteResult> CreateOwnIdpOwnCompanyUser([FromBody] Us
     /// <response code="200">Returns the client roles.</response>
     /// <response code="400">The language does not exist.</response>
     /// <response code="404">The app was not found.</response>
+    [Obsolete("to be replaced by endpoint /user/owncompany/roles/apps/{appid}. Remove as soon frontend is adjusted")]
     [HttpGet]
     [Authorize(Roles = "view_client_roles")]
     [Route("app/{appId}/roles")]
-    [ProducesResponseType(typeof(IAsyncEnumerable<ClientRoles>), StatusCodes.Status200OK)]
+    [ProducesResponseType(typeof(IEnumerable<ClientRoles>), StatusCodes.Status200OK)]
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/Administration.Service/Controllers/UserController.cs:364]: from my understanding this annotation doesn't change the behavior of the endpoint, it defines the swagger.json. So it's just a question what looks better - technically it's equivalent.

File: src/portalbackend/PortalBackend.PortalEntities/Entities/CompanyRole.cs [hunk 2]
```diff
@@ -45,6 +45,8 @@ public CompanyRole(CompanyRoleId companyRoleId) : this()
     public string Label { get; set; }
 
     // Navigation properties
+    public virtual CompanyRoleAssignedRoleCollection? CompanyRoleAssignedRoleCollection { get; set; }
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/portalbackend/PortalBackend.PortalEntities/Entities/CompanyRole.cs:48]: actually not 'could' but 'should' ;-)
- done -

File: src/portalbackend/PortalBackend.PortalEntities/Entities/CompanyRoleRegistrationData.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        IsRegistrationRole = isRegistrationRole;
+    }
+
+    [Key]
+    public CompanyRoleId CompanyRoleId { get; private set; }
+
+    public bool IsRegistrationRole { get; private set; }
+
+    // Navigation properties
+    public virtual CompanyRole CompanyRole { get; set; } = null!;
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/portalbackend/PortalBackend.PortalEntities/Entities/CompanyRoleRegistrationData.cs:40]: done

File: src/portalbackend/PortalBackend.PortalEntities/Entities/UserRoleAssignedCollection.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        UserRoleId = userRoleId;
+        UserRoleCollectionId = userRoleCollectionId;
+    }
+
+    public Guid UserRoleId { get; private set; }
+    public Guid UserRoleCollectionId { get; private set; }
+
+    // Navigation properties
+    public virtual UserRoleCollection UserRoleCollection { get; private set; } = null!;
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/portalbackend/PortalBackend.PortalEntities/Entities/UserRoleAssignedCollection.cs:35]: done

File: src/administration/Administration.Service/BusinessLogic/UserBusinessLogic.cs [hunk 5]
```diff
@@ -82,12 +82,16 @@ public async IAsyncEnumerable<string> CreateOwnCompanyUsersAsync(IEnumerable<Use
...
+        var roleDatas = await GetOwnCompanyUserRoleData(distinctRoles, iamUserId).ConfigureAwait(false);
+
         var userCreationInfoIdps = userList.Select(user =>
-            new UserCreationInfoIdp(
+            new UserCreationRoleDataIdpInfo(
                 user.firstName ?? "",
                 user.lastName ?? "",
                 user.eMail,
-                user.Roles,
+                roleDatas?.IntersectBy(user.Roles, roleData => roleData.UserRoleText) ?? Enumerable.Empty<UserRoleData>(),
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/Administration.Service/BusinessLogic/UserBusinessLogic.cs:94]: done

File: src/administration/Administration.Service/Controllers/UserController.cs [hunk 6]
```diff
@@ -200,6 +203,46 @@ public async Task<CreatedAtRouteResult> CreateOwnIdpOwnCompanyUser([FromBody] Us
...
+    /// <response code="200">Roles got successfully updated user account.</response>
+    /// <response code="400">Invalid User roles for client</response>
+    /// <response code="404">User not found</response>
+    [HttpPut]
+    [Authorize(Roles = "modify_user_account")]
+    [Route("owncompany/users/{companyUserId}/apps/{appId}/roles")]
+    [ProducesResponseType(typeof(IEnumerable<UserRoleWithId>), StatusCodes.Status200OK)]
+    [ProducesResponseType(typeof(ErrorResponse), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ErrorResponse), StatusCodes.Status404NotFound)]
+    public Task<IEnumerable<UserRoleWithId>> ModifyAppUserRolesAsync([FromRoute] Guid companyUserId, [FromRoute] Guid appId, [FromBody] IEnumerable<string> roles) =>
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/Administration.Service/Controllers/UserController.cs:243]: service-offerings cannot have roles that would be assignable in keycloak as service-offerings do not have an associated appinstance. And core-offers have their own endpoint as they are not linked to the app-instance via subscriptions.
If this endpoint is being called with the offerId of a core or service-offering no iamClientId would be found and the businesslogic would throw NotFoundException.

File: src/administration/Administration.Service/BusinessLogic/UserRolesBusinessLogic.cs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    private readonly IProvisioningManager _provisioningManager;
+
+    public UserRolesBusinessLogic(IPortalRepositories portalRepositories, IProvisioningManager provisioningManager)
+    {
+        _portalRepositories = portalRepositories;
+        _provisioningManager = provisioningManager;
+    }
+
+    public IAsyncEnumerable<OfferRoleInfos> GetCoreOfferRoles(string iamUserId, string? languageShortName) =>
+        _portalRepositories.GetInstance<IUserRolesRepository>().GetCoreOfferRolesAsync(iamUserId, languageShortName == null ? IUserRolesBusinessLogic.DEFAULT_LANGUAGE : languageShortName)
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/Administration.Service/BusinessLogic/UserRolesBusinessLogic.cs:44]: done.

File: src/administration/Administration.Service/BusinessLogic/UserRolesBusinessLogic.cs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            .PreSortedGroupBy(x => x.OfferId)
+            .Select(x => new OfferRoleInfos(x.Key, x.Select(s => new OfferRoleInfo(s.RoleId,s.RoleText,s.Description))));
+
+    public IAsyncEnumerable<OfferRoleInfo> GetAppRolesAsync(Guid appId, string iamUserId, string? languageShortName) =>
+        _portalRepositories.GetInstance<IUserRolesRepository>()
+            .GetAppRolesAsync(appId, iamUserId, languageShortName == null ? IUserRolesBusinessLogic.DEFAULT_LANGUAGE : languageShortName);
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/Administration.Service/BusinessLogic/UserRolesBusinessLogic.cs:50]: done

File: src/portalbackend/PortalBackend.PortalEntities/Entities/CompanyRole.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/PortalBackend.PortalEntities/Entities/CompanyRole.cs:48]: agree, we maybe have to change a few of our navigation properties in the future.

File: src/administration/Administration.Service/Controllers/UserController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Phil91) [reply src/administration/Administration.Service/Controllers/UserController.cs]: you're right. shouldn't make a huge difference.

File: src/administration/Administration.Service/Controllers/UserController.cs [hunk 9]
```diff
@@ -280,6 +323,30 @@ public async Task<CreatedAtRouteResult> CreateOwnIdpOwnCompanyUser([FromBody] Us
...
+    /// <returns>Returns the client roles for the given app.</returns>
+    /// <remarks>Example: GET: api/administration/user/owncompany/app/D3B1ECA2-6148-4008-9E6C-C1C2AEA5C645/roles</remarks>
+    /// <response code="200">Returns the client roles.</response>
+    /// <response code="400">The language does not exist.</response>
+    /// <response code="404">The app was not found.</response>
+    [HttpGet]
+    [Authorize(Roles = "view_client_roles")]
+    [Route("owncompany/roles/apps/{appId}")]
+    [ProducesResponseType(typeof(OfferRoleInfos), StatusCodes.Status200OK)]
+    public IAsyncEnumerable<OfferRoleInfo> GetAppRolesAsync([FromRoute] Guid appId, [FromQuery] string? languageShortName = null) =>
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/administration/Administration.Service/Controllers/UserController.cs:347]: thanks for the explanation

**Review by Phil91 — APPROVED** [review]

**Pull request merged**